File: JAVA/tcp-cli-srv/TcpCliSoma.java
import java.io.*; 
import java.net.*; 

class TcpCliSoma
{    
static InetAddress IPdestino;
static Socket sock;

public static void main(String args[]) throws Exception    
{       
if(args.length!=1) {
	System.out.println("Deve indicar o host de destino na linha de comando");
	System.exit(1); }

try { IPdestino = InetAddress.getByName(args[0]); }
catch(UnknownHostException ex) {
	System.out.println("O nome de host de destino fornecido (" + args[0] + ") nao foi resolvido");
	System.exit(1); }

try { sock = new Socket(IPdestino, 9999); }
catch(IOException ex) {
	System.out.println("A ligacao TCP falhou.");
	System.exit(1); }

BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
DataOutputStream sOut = new DataOutputStream(sock.getOutputStream());
DataInputStream sIn = new DataInputStream(sock.getInputStream());


String frase;
long f,i,n,num;
do
   {
   do
	{
	num=-1;
	while(num<0) {
	  System.out.print("Numero inteiro positivo a somar (zero para terminar): ");
	  frase = in.readLine();
	  try { num=Integer.parseInt(frase); } catch(NumberFormatException ex) {num=-1;}
	  if(num<0) System.out.println("Formato nao aceite");
	  }
	n=num; for(i=0;i<4;i++) {sOut.write((byte)(n%256)); n=n/256; }
	}
   while(num!=0);
   num=0; f=1; for(i=0;i<4;i++) {num=num+f*sIn.read(); f=f*256;}
   System.out.println("A soma recebida foi: " + num);
   }
while(num!=0);
sock.close();
} 
}

File: C/tcp-cli-srv/tcp_cli_soma.c
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>


// ler do stdin uma frase com menos de S carateres para o buffer B
#define GETS(B,S) {fgets(B,S-2,stdin);B[strlen(B)-1]=0;}

void main(int argc, char **argv)
{
struct sockaddr_in target;
int sock;
unsigned long f, i, n, num;
unsigned char bt, linha[30];
   
if(argc!=2) {printf("Utilização: %s Endereço-IP-Servidor\n",argv[0]);exit(1);}

sock=socket(AF_INET,SOCK_STREAM,0);

bzero((char *)&target,sizeof(target));
target.sin_family = AF_INET;
target.sin_addr.s_addr=inet_addr(argv[1]);
target.sin_port=htons(9999);

if(connect(sock,(struct sockaddr *)&target, sizeof(target))==-1) 
	{puts("A ligação falhou"); exit(1);}

do 
	{
        do 
        	{
		printf("Numero inteiro positivo a somar (zero para terminar): ");
		GETS(linha,30);
		while(sscanf(linha,"%li",&num)!=1 || num<0)
			{
			puts("Numero invalido");
			GETS(linha,30);
			}
		n=num;
		for(i=0;i<4;i++) 
			{bt=n%256; write(sock,&bt,1); n=n/256; }
        	}
        while(num);
	num=0; f=1; for(i=0;i<4;i++) {read(sock,&bt,1); num=num+bt*f; f=f*256;}
        printf("SOMA=%lu\n",num);
        }
    while(num);
close(sock);
}

